### PR TITLE
fix(deep-cody): Rate limiter synchronization issue

### PR DESCRIPTION
**Pull request by abeatrix**
Follow up on https://github.com/sourcegraph/cody/pull/6366

This PR fixes several issues with the Deep Cody rate limiter implementation:

Root Cause Fix: Rate Limit State Synchronization

- Fixed race condition in LocalStorageProvider.ts where concurrent Promise.all() usage led to inconsistent state between quota and timestamp storage
- Changed to sequential storage operations to ensure timestamp is always saved after quota update
- This resolves the core issue where users were locked out due to incorrect time calculations (CODY-4529)

## Test plan

- Added precise tests for time-based quota calculations
- Enhanced test coverage for various timing scenarios

Before this change, the Deep Cody rate limit error would always return the "Retry time" with current time + 24 hours to indicate the current time is being used as the last used time.

Submitted at 6:44

image

Submitted again at 6:45

image

After this change, the "Retry Time" should stick with the last used time until it's been reset after 24 hours.

### CODE REVIEW

File: vscode/src/services/LocalStorageProvider.ts [hunk 1]
```diff
@@ -357,46 +357,8 @@ class LocalStorage implements LocalStorageForModelPreferences {
     }
 
     public async setDeepCodyUsage(newQuota: number, lastUsed: string): Promise<void> {
-        await Promise.all([
-            localStorage.set(localStorage.keys.deepCodyDailyUsageCount, newQuota - 1),
-            localStorage.set(localStorage.keys.deepCodyLastUsedDate, lastUsed),
-        ])
-    }
-
-    public isAtDeepCodyDailyLimit(DAILY_QUOTA?: number): string | undefined {
```
Comment (abeatrix) [inline vscode/src/services/LocalStorageProvider.ts:366]: this was left over that is not used.

File: vscode/src/chat/agentic/DeepCodyRateLimiter.ts [hunk 2]
```diff
@@ -28,13 +28,6 @@ export class DeepCodyRateLimiter {
         const lastUsedTime = lastUsed.getTime()
         const timeDiff = now - lastUsedTime
 
-        // Reset quota if more than 24 hours have passed
-        if (timeDiff >= this.ONE_DAY_MS) {
```
Comment (abeatrix) [inline vscode/src/chat/agentic/DeepCodyRateLimiter.ts:32]: removing the unnecessary change from last PR.

File: vscode/src/services/LocalStorageProvider.ts [hunk 3]
```diff
@@ -357,46 +357,8 @@ class LocalStorage implements LocalStorageForModelPreferences {
     }
 
     public async setDeepCodyUsage(newQuota: number, lastUsed: string): Promise<void> {
-        await Promise.all([
-            localStorage.set(localStorage.keys.deepCodyDailyUsageCount, newQuota - 1),
```
Comment (abeatrix) [inline vscode/src/services/LocalStorageProvider.ts:361]: newQuota - 1  was causing the quota to be deducted twice.

**Review by dominiccooney — APPROVED** [review]
This is an improvement and I'm glad we fixed the real issue.

I didn't see the PR that had isAtLimit return the time to wait, however some feedback now looking at the whole thing. You can fix this in a follow up change to keep this change simple to backport:

- It is bad to have `isAtLimit` decrement the quota because it *sounds* like this method just tests something. But it also consumes the limit. In a follow up, consider renaming this to make it obvious that it has a side effect of also consuming the limit. (You saw the problem yourself... we decremented the limit twice and it wasn't obvious we were doing so.)
- The `isAtLimit` naming gives no hint that it returns a time in seconds to wait for quota.
- The `isAtLimit` *types* are odd: Seconds to wait is a `number`, why are we returning it as a `string`?
- The "time to wait" calculation looks wrong. For example, if you used your quota, wait an hour, then chat this calculation says to wait 23 hours. But actually you get your next token in 12 minutes. (There are 20 tokens/day = 1 token every 72 minutes, and you already waited 60 minutes. Now, maybe the UX is better if people back up a few tokens so they can do more than one message at a time. But we probably don't need to tell them to wait 24 hours...)
